### PR TITLE
Allow usernames to include more characters when using the `!username` command

### LINKED ISSUE #1359: Can't store username that contains underscore
**Describe the bug**
I tried to store my username on Libera Chat (Logan_) for SASL authentication using `!username Logan_`, but I got the message that it contains invalid characters not supported by IRC:
```
!username Logan_

Username contained invalid characters not supported by IRC.
```

This message is definitely incorrect because my username is registered with Libera Chat with an underscore. It looks like [this is because `_` is used as a delimiter by the bridge](https://github.com/matrix-org/matrix-appservice-irc/blob/develop/src/irc/IdentGenerator.ts#L226). Does this mean I can't authenticate my Libera username using Matrix?

**To Reproduce**
1. Open a DM with the bridge user
2. Say `!username` and then a username containing an underscore
3. See error

**Expected behavior**
I expected to be able to store my IRC username with Matrix and automatically authenticate using SASL.

**Desktop (please complete the following information):**
 - OS: Ubuntu
 - Browser Firefox
 - Version 90.0a1

### PR DESCRIPTION
**Pull request by progval**
Fixes https://github.com/matrix-org/matrix-appservice-irc/issues/1359
Supercedes https://github.com/matrix-org/matrix-appservice-irc/pull/1367 and https://github.com/matrix-org/matrix-appservice-irc/pull/1538, which were both too restrictive

This change allows for a much more permissive username when users are setting their own username.

### CODE REVIEW

**Comment by progval** [comment]
Signed-off-by: Valentin Lorentz \<progval plus matrixsignoff at progval dot net\>

File: src/bridge/AdminRoomHandler.ts [hunk 1]
```diff
@@ -39,6 +39,15 @@ enum CommandPermission {
...
 const SANE_USERNAME_LENGTH = 64;
 
+// Technically, anything but \0 is allowed as username (aka. authcid and authzid):
+// https://www.rfc-editor.org/rfc/rfc4616#section-2
+//
+// However, IRC services are very unlikely to allow the username to contain CR or LF
+// (because they would not fit in the wire format) or spaces (because usernames are
+// usually followed by passwords in "PRIVMSG NickServ :REGISTER" commands and
+// IRCv3 draft/account-registration.
+const SASL_USERNAME_INVALID_CHARS_PATTERN = /[ \0\r\n]+/;
```
Comment (Half-Shot) [inline src/bridge/AdminRoomHandler.ts]: Wondering if we at least limit to printable characters. I can't see the value in allowing most of these.

```suggestion
const SASL_USERNAME_INVALID_CHARS_PATTERN = /[\0-\38\7F]+/;
```

File: src/bridge/AdminRoomHandler.ts [hunk 2]
```diff
@@ -503,10 +513,10 @@ export class AdminRoomHandler {
                     `Username is longer than the maximum permitted by the bridge (${SANE_USERNAME_LENGTH}).`
                 );
             }
-            else if (IdentGenerator.sanitiseUsername(username) !== username) {
+            else if (invalidChars !== null) {
                 notice = new MatrixAction(
                     ActionType.Notice,
-                    `Username contained invalid characters not supported by IRC.`
+                    `Username contained invalid characters not supported by IRC (${JSON.stringify(invalidChars.join(""))}).`
```
Comment (Half-Shot) [inline src/bridge/AdminRoomHandler.ts]: Wonder if we should make the unprintable, printable?

**Review by Half-Shot — COMMENTED** [review]
Looking really nice, just a couple of nits.

File: src/bridge/AdminRoomHandler.ts (hunk 1 above)
Comment (progval) [reply src/bridge/AdminRoomHandler.ts]: I think you mean `[\0-\40\x7F]` (38 isn't valid octal, the octal for space is 40, and 7F is the hex code for DEL, not the octal code).

File: src/bridge/AdminRoomHandler.ts (hunk 2 above)
Comment (progval) [reply src/bridge/AdminRoomHandler.ts]: Doesn't `JSON.stringify` do it?

File: src/bridge/AdminRoomHandler.ts (hunk 1 above)
Comment (Half-Shot) [reply src/bridge/AdminRoomHandler.ts]: yeah, bad copy paste. That one ^

File: src/bridge/AdminRoomHandler.ts (hunk 2 above)
Comment (Half-Shot) [reply src/bridge/AdminRoomHandler.ts]: So it does!

**Comment by Half-Shot** [comment]
Assuming tests pass this will be good enough for a merge.

**Review by Half-Shot — APPROVED** [review]

**Comment by Half-Shot** [comment]
I'm happy with the contrib requirements and I believe docker builds are broken for non-collaborators atm (due to needing a login). @progval if you can make the lint happy I'll chuck this in the release tomorrow.

**Comment by Half-Shot** [comment]
:star: let's go!

**Pull request merged**